File: chapters/software-stack/high-level-languages/drills/tasks/high-level-lang/solution/src/copy.go
package main

/* REMOVE 5 */
import (
	"io"
	"os"
)

func main() {
	/* TODO 24: Copy file from command line arguments */
	if len(os.Args) < 3 {
		os.Exit(1)
	}

	fin, err := os.Open(os.Args[1])
	if err != nil {
		os.Exit(1)
	}

	fout, err := os.Create(os.Args[2])
	if err != nil {
		fin.Close()
		os.Exit(1)
	}

	_, err = io.Copy(fout, fin)
	if err != nil {
		fin.Close()
		fout.Close()
		os.Exit(1)
	}

	fin.Close()
	fout.Close()
}

File: chapters/software-stack/high-level-languages/drills/tasks/high-level-lang/solution/src/copy.c
#include <stdio.h>

int main(int argc, char* argv[])
{
	char buf[128];
	FILE *fin, *fout;

	/* TODO 28: Copy file from command line arguments */
	if (argc < 3)
		return 1;

	fin = fopen(argv[1], "rb");
	if (!fin)
		return 1;

	fout = fopen(argv[2], "wb");
	if (!fout) {
		fclose(fin);
		return 1;
	}

	while (!feof(fin) && !ferror(fin)) {
		size_t bytes = fread(buf, 1, sizeof buf, fin);
		fwrite(buf, 1, bytes, fout);
	}

	if (!ferror(fin) && !ferror(fout)) {
		fclose(fin);
		fclose(fout);
		return 0;
	}

	fclose(fin);
	fclose(fout);

	return 1;
}
